flink-runtime/src/main/java/org/apache/flink/runtime/operators/hash/ReduceHashTable.java

```diff
@@ -0,0 +1,1014 @@
+				final long firstPointer = bucketSegment.getLong(bucketOffset);
+
+				long ptrToAppended = recordArea.noSeekAppendPointerAndRecord(firstPointer, record);
+				bucketSegment.putLong(bucketOffset, ptrToAppended);
+			}
+			recordArea.freeSegmentsAfterAppendPosition();
+			holes = 0;
+
+		} catch (EOFException ex) {
+			throw new RuntimeException("Bug in ReduceHashTable: we shouldn't get out of memory during a rebuild, " +
+				"because we aren't allocating any new memory.");
+		}
+	}
+
+	/**
+	 * Emits all elements currently held by the table to the collector,
+	 * and resets the table. The table will have the same number of buckets
+	 * as before the reset, to avoid doing resizes again.
+	 */
+	public void emitAndReset() throws IOException {
+		final int oldNumBucketSegments = bucketSegments.length;
+		emit();
+		close();
+		open(oldNumBucketSegments);
+	}
+
+	/**
+	 * Emits all elements currently held by the table to the collector.
+	 */
+	public void emit() throws IOException {
+		T record = buildSideSerializer.createInstance();
+		EntryIterator iter = getEntryIterator();
+		while ((record = iter.next(record)) != null) {
+			outputCollector.collect(record);
+			if (!objectReuseEnabled) {
+				record = buildSideSerializer.createInstance();
+			}
+		}
+	}
+
+	/**
+	 * If there is wasted space due to updates records not fitting in their old places, then do a compaction.
+	 * Else, throw EOFException to indicate that memory ran out.
+	 * @throws IOException
+	 */
+	private void compactOrThrow() throws IOException {
+		if (holes > 0) {
+			rebuild();
+		} else {
+			throw new EOFException("ReduceHashTable memory ran out. " + getMemoryConsumptionString());
+		}
+	}
+
+	/**
+	 * @return String containing a summary of the memory consumption for error messages
+	 */
+	private String getMemoryConsumptionString() {
+		return "ReduceHashTable memory stats:\n" +
+			"Total memory:     " + numAllMemorySegments * segmentSize + "\n" +
+			"Bucket area:      " + numBuckets * 8  + "\n" +
+			"Record area:      " + recordArea.getTotalSize() + "\n" +
+			"Staging area:     " + stagingSegments.size() * segmentSize + "\n" +
+			"Num of elements:  " + numElements + "\n" +
+			"Holes total size: " + holes;
+	}
+
+	/**
+	 * This class encapsulates the memory segments that belong to the record area. It
+	 *  - can append a record
+	 *  - can overwrite a record at an arbitrary position (WARNING: the new record must have the same size
+	 *    as the old one)
+	 *  - can be rewritten by calling resetAppendPosition
+	 *  - takes memory from ReduceHashTable.freeMemorySegments on append
+	 */
+	private final class RecordArea
+	{
+		private final ArrayList<MemorySegment> segments = new ArrayList<>();
+
+		private final OutputView outView;
+		private final RandomAccessInputView inView;
+
+		private final int segmentSizeBits;
+		private final int segmentSizeMask;
+
+		private long appendPosition = 0;
+
+
+		private final class OutputView extends AbstractPagedOutputView {
```

Comment:
I tried this and now I remember why I did it this way: I need random access, and `SimpleCollectingOutputView` doesn't have that.

`seekOutput` is overridden to make it public. I now that generally it is not a good practice to override a protected method just to make it public, but I think that this is the best solution here. The alternatives would be to
- Implement a combination of `SimpleCollectingOutputView` and `RandomAccessOutputView`: this seems to unnecessarily complicate things
- Make `RecordArea` itself extend `OutputView`: I think this would obscure that structure of the `RecordArea` that it has an outView and an inView over a single `ArrayList` of segments.